### ISSUE TITLE
Cache stream for each user in background for faster stream loading

### ISSUE THREAD

**Issue by jaywink** [diaspora/diaspora#5984]
*Logging this thing that has been in my head for a while, for future discussion and implementation if wanted.*

Currently we always look at which posts should be shown when a user loads the stream. While being ok, it is an expensive operation to perform.

To provide faster near instant stream loading (like Facebook), it could be a good idea to cache the latest stream for each user in the background. This would move the expensive operations to the background processing and allow when loading the stream to just fetch a readily calculated list of post id's for rendering. It could even be a list maintained in Redis as a per user queue if that would be better for performance.

So, to clarify, what we do now is:

* User loads stream.
* Application asks db the latest posts for that should be included in the stream by looking at the users the user follows, the tags the user follows and also takes into consideration any ignoring etc. (disclaimer: I haven't looked at the code but afaik it goes along these lines).
* Posts are pushed to frontend and rendered.

This could be more efficient, though of course more complex too:

* User loads stream.
* Applications fetches a subset of the pre-generated post id's.
* Posts are pushed to frontend and rendered.

When a new post is written, the following happens now:

* Post is saved
* Post is federated to remote pods

For the caching the following should happen

* Post is saved
* Post is federated to remote pods
* Application creates a sidekiq queue to refresh the stream of local users. When this is executed, the app will push the post id to any pre-calculated user stream lists which should see the post, following the earlier logic in reverse.

For this feature to be effective (because a delay is introduced otherwise depending on sidekiq queue size), real time connectivity would be needed between front-end clients and the back-end. After this would be implemented, the "pushing to stream lists" could also check whether a client is connected currently and if yes, push the post directly to the user front-end client for rendering.

**Comment by Raven24** [comment]
I've had a similar idea lurking around in my brain, too, for quite a while. (yours is a bit more thought-out, though)
I'd be very much in favour, even if the "cache" would just consist of the post ids to make the db query with, we should see a tremendous speed-up.

**Comment by Flaburgan** [comment]
Why don't we do that client side? The stream is loaded with Javascript, so we could just get the new posts with AJAX instead of reloading everything. Like, sending a request saying "the most recent post I have in my stream is this one, please send me every new posts since then". This would also allow an "auto-refresh" feature in the future.

**Comment by jaywink** [comment]
@Raven24 

> I'd be very much in favour, even if the "cache" would just consist of the post ids to make the db query with, we should see a tremendous speed-up.

Yeah I actually meant just saving the post id's and then querying the stream with those - not caching the post content as that would be duplicating information and would kill any post edit feature for sure :)

@Flaburgan , yeah as said, client side instant updates would be needed too in reality, since otherwise the cached stream could become stale if sidekiq is too busy to stay up with jobs. It happens on Facebook too sometimes - you reload and get really old posts and then 1s later newer stuff is populated on the top.

I personally would object to auto-refreshing the stream, that would be really confusing UX. But auto-loading them on the top and then a reveal button would work. Tweetdeck works really nicely like this, the reveal button with "X new tweets" hovers on top when there is something new, clicking it scrolls to top.

But this is really a separate thing from the caching of stream which I don't see how it could be done in the front-end, if that is what you meant.

**Comment by CSammy** [comment]
I have thought about this a bit myself, and for newly arising requests of implementing something like this, I am going to detail a few of those thoughts.

There are a few possibilities on how to make this work. I will give examples with numbers from joindiaspora.org as reference pod.

1. Store these IDs for all posts and users for the entire stream.

   - **Advantages**
      - One select, after which all IDs are known, no difficult joins when loading the stream.
   - **Disadvantages**
      - In order to show the relation, user ID and post ID columns are needed. To be able to select the latest *k* posts, a rank is needed. Since posts may arrive late, it is better to use a time-based type instead of a running number. For MySQL, this amounts to 16 bytes per row. Assuming every user has on average 1000 posts in their stream, on the reference pod this amounts to `340000 * 1000 * 16 bytes = ~5GiB`. For a cache for one single stream out of all the stream types we offer, this is a bit big, but not undoable. Please keep in mind that indices will need to be added (most likely a composite index, user ID + timestamp).
      - For every incoming post, the various possibilities how any user might be able to see this post need to be checked. Namely, it needs to be checked who is following this user, and for every tag of the post all users who follow this post need to be retrieved. In addition, for every of those users it needs to be checked whether they ignore the author. I have probably forgotten a few constraints. *Basically, the stream is calculated for all users on arrival of every post.* Especially with larger user bases, this becomes a huge constant workload in contrast to the workload caused by actually active users.
      - For every incoming post, up to -- depending on tags and visibility -- several thousand entries need to be written to this table and corresponding indices. This puts a lot of strain on the disk, making it age sooner.

2. Store the IDs only for a certain time span.
   - **Disadvantages**
      - All of variant 1, except the storage requirements are linear to user base only instead of both user and post count.
      - Outdated entries need to be removed, causing more load on the server and strain on the disk.
      - A second stream building logic covering the case where this cache table is a miss needs to be maintained.

3. Store only a certain amount of IDs per user.
   - **Disadvantages**
      - All of variant 2.
      - In addition, the amount of IDs in cache need to be tracked per user.

4. Store the IDs only for active users.
   - **Advantages**
      - Storage requirements go down in comparison to variant 1.
   - **Disadvantages**
      - The "active" status of every user needs to be calculated on every post arrival, or this metric needs to be tracked additionally and checked when calculating new entries to the cache table
      - The cache table needs to be kept up to date depending on users becoming active or inactive.
      - A second stream building logic is still necessary.
      - As of this writing, no simple fast method of only checking active users without looking at all users comes to mind other than partitioning into two user bases (active and inactive). This would add a lot of additional complexity.

To me, the bottom line is that it either does not scale or will become so complex that we're better off just leaving it like this, because nobody will be able to maintain this mess consistently even if the developers are able to pull it off.

> There are only two hard things in Computer Science: cache invalidation and naming things.
~ [presumably by Phil Karlton](https://twitter.com/timbray/status/506146595650699264)

**Comment by jaywink** [comment]
Ha, didn't remember I had logged this for diaspora years back :P I've implemented this same in Socialhome to ensure streams always load in constant (near instant) speed, not depending on stream logic, users, content amount, etc. I didn't want to constantly see a loading spinner like I used to on my Diaspora pod. It was a bit more work than I anticipated but it's totally worth it 👍 Regarding the above, I chose the following to keep the feature light on storage:

* The amount of ID's configurable by admin, by default set to 100 per user per stream, which is very reasonable on Redis.
* Future expansion is to add an extra check on calculating only for users who have been active in X months.

For storage I chose Redis since it's super fast to read from and super easy to trim to size. Losing the data also isn't a catastrophe since if there are no precalculated ID's the stream generation just falls back to ORM.

> There are only two hard things in Computer Science: cache invalidation and naming things.

Seeing something as hard doesn't *always* mean it should be avoided ;) Think of society if we never took the hard route. Caching is something that should not be done lightly, totally agree to that.

Just sharing some thoughts, and @CSammy wanted to move [discussion](https://jasonrobinson.me/content/1132788/got-a-clone-of-my-pod-running-on-on-google-kuber/) to github, so replied here. My code is [here](https://github.com/jaywink/socialhome/blob/master/socialhome/streams/streams.py) should anyone be interested.
